Cm/pse 6 rework (#26)
Ada-Developers-Academy/core-pses: 06/merging-sorted-lists.md

| List 1                | List 2       | Output                            |
| --------------------- | ------------ | --------------------------------- |
| [1, 2, 4, 5]          | [6]          | [1, 2, 4, 5, 6]                   |
| [-30, -10, 0, 15, 16] | [-20, -5, 5] | [-30, -20, -10, -5, 0, 5, 15, 16] |

Sourced from: [Leetcode](https://leetcode.com/problems/merge-sorted-array/)

## Prompts

<!-- Question 1 -->
<!-- prettier-ignore-start -->
### !challenge
* type: paragraph
* id: ce61e622-d172-495c-a6b2-653f0c3c6624
* title: Ask Clarifying Questions
* points: 3
* topics: pse
##### !question

List three or more questions whose answers would clarify the problem statement

##### !end-question
##### !hint

Consider the following for inspiration:

- [About PSEs](../about-pses/about-pses.md)
- [Our example PSE with example answers](../about-pses/example-pse.md)
- Any past PSEs you may have

##### !end-hint
##### !explanation

Here are some example clarifying questions:

> - What about empty lists

Reviewer: For consistency
```suggestion
- What about empty lists?
```

> - What about empty lists
> - What to do about elements of equal value? for example `[3]` and `[3]`, do both go in result?
> - Can numbers be negative?

Reviewer 2: Minor suggestion to rephrase with complete sentence:

```suggestion
- How does the function handle empty lists?
- How does the function sort elements of equal value? 
    - For example: `list1 = [3]` and `list2 = [3]`. Are both values include in returned list?
- Can numbers in the input lists be negative?
```

    # Assert
    assert result == []

def test_list_1_all_before_list_2():
    # alternative test case
    # Arrange
    list1 = [-50, -25, 0]
    list2 = [10, 20, 30]

    # Act
    result = merge_lists(list1, list2)

    # Assert
    assert result == [-50, -25, 0, 10, 20, 30]

```

##### !end-explanation
### !end-challenge
<!-- prettier-ignore-end -->
<!-- Question 3 -->
<!-- prettier-ignore-start -->
### !challenge
* type: paragraph
* id: 1f6d0c1e-df66-454b-a6e8-61ce8726b859
* title: Create Logical Steps
* points: 3
* topics: pse
##### !question

Without writing code, describe how you would implement `merge_lists` in enough detail that someone else could write the code. 

* It may be helpful to break up the problem/algorithm into smaller subproblems/algorithms. For example, 1. Handle invalid input, 2. Given valid input, perform the computation/solve the problem/etc.
* Your logical steps could take the form of a numbered list, pseudo code, or anywhere in between. What's important at this stage is to think through and outline the implementation before writing code.

##### !end-question

##### !placeholder

Write the logical steps here.

##### !end-placeholder

### !explanation

> Example Steps for an O(n<sup>2</sup>) solution:
> 
> 1. Check for valid input
> 1. set i to 0 and j to 0
> 1. create result as an empty list
> 1. while i < length of list 1 and j < length of list 2
>     - if list[i] < list[j]
>         - append list[i] to result
>         - increment i
>     - else
>         - append list[j] to result
>         - increment j
> 1. while i < length of list 1
>     - append list[i] to result
>     - increment i
> 1. while j < length of list 2
>     - append list[j] to result
>     - increment j
> 1. return result
> 

Reviewer 2: Minor suggestions to be consistent with previous PSEs and update list to `list1` or `list2`

```suggestion
Example steps for an O(n<sup>2</sup>) solution:

1. Check for valid input
1. set `i` to 0 and `j` to 0
1. initialize `result` as an empty list
1. while `i` < length of `list1` and `j` < length of `list2`
    - if `list1[i]` < `list2[j]`
        - append `list1[i]` to `result`
        - increment `i`
    - else
        - append `list2[j]` to `result`
        - increment `j`
1. while `i` < length of `list1`
    - append `list1[i]` to `result`
    - increment `i`
1. while `j `< length of `list2`
    - append `list2[j]` to `result`
    - increment `j`
1. return `result`

```
